### ISSUE TITLE
Non-consecutive text nodes are errorneously merged, locations wrongly reported

### ISSUE THREAD

**Issue by squidfunk** [inikulin/parse5#282]
## TL;DR: the parser currently merges non-consecutive text tokens

I encountered this problem when enabling source location information - the locations are currently reported wrong for nodes and text nodes. I started digging for two hours, those are my insights.

### Source

It's reproducible with the latest installable version and the following HTML:

``` html
<html>
  <head></head>
  <body>
    Foobar
  </body>
      </html>



```

I intentionally added the odd whitespace for clarity. The newlines at the end of the file are also intentional.

### Tokens

Parsing this source reveals the following tokens:

``` js
{ nodeName: 'body',
  tagName: 'body',
  attrs: [],
  namespaceURI: 'http://www.w3.org/1999/xhtml',
  childNodes: [Array],
  parentNode: null, // removed for clarity,
  sourceCodeLocation: 
   { startLine: 3,
     startCol: 3,
     startOffset: 25,
     endLine: 9,
     endCol: 1,
     endOffset: 69,
     startTag: 
      { startLine: 3,
        startCol: 3,
        startOffset: 25,
        endLine: 3,
        endCol: 9,
        endOffset: 31 },
     endTag: 
      { startLine: 5,
        startCol: 3,
        startOffset: 45,
        endLine: 5,
        endCol: 10,
        endOffset: 52 } } }
```

According to the source, the `body` should start at `[3,3]` and end at `[5,10]`. This is correctly reported for the start and closing tags, but not for the overall location. The location is reported as `[3,3]` to `[9,1]`, so the location parser swallowed more than it should, actually until `EOF`.

Now the text token:

``` js
{ nodeName: '#text',
  value: '\n    Foobar\n  \n      \n\n\n',
  parentNode: 
   { nodeName: 'body',
     tagName: 'body',
     attrs: [],
     namespaceURI: 'http://www.w3.org/1999/xhtml',
     childNodes: [ [Circular] ],
     parentNode: null, // removed for clarity
     sourceCodeLocation: 
      { startLine: 3,
        startCol: 3,
        startOffset: 25,
        endLine: 9,
        endCol: 1,
        endOffset: 69,
        startTag: [Object],
        endTag: [Object] } },
  sourceCodeLocation: 
   { startLine: 3,
     startCol: 9,
     startOffset: 31,
     endLine: 9,
     endCol: 1,
     endOffset: 69 } }

```

This is even stranger. Actually, five text tokens should be reported, because there is (1) whitespace before the `html` tag, (2) whitespace before the `body`, (3) the "Foobar" text including whitespace and (4) the whitespace before the `html` closing tag, and (5) the whitespace after that tag.

As one can see, only one token, namely `'\n    Foobar\n  \n      \n\n\n''` is reported which includes all the text from the file where five should have been reported. This does also happen if the source code location is turned off, so the error seems to originate from the parser.

I actually have no idea where to continue debugging. Maybe someone with some knowledge of the code base cares to comment. I hope that I provided enough information and my problem is easily reproducible. If something is missing, please let me know.

**Version: 5.1.0**

**Comment by RReverser** [comment]
What you're calling tokens here is actually nodes (DOM-like), and for nodes indeed non-consecutive items can be reordered and merged as per spec.

Whenever in doubt, you might want to try same HTML in any browser and inspect the DOM - you should see the same behaviour as in parse5. If not, then it's indeed a bug in one or another.

**Comment by squidfunk** [comment]
Oh okay, I wasn't aware of that. Thanks for pointing it out!

**Comment by meisl** [comment]
I think I've found a related issue, but actually with the location info in tokens from the tokenizer, ie. a *real* issue.

It seems it's got to do with coercing character tokens of the same type (`Tokenizer.CHARACTER_TOKEN`, `Tokenizer.WHITESPACE_CHARACTER_TOKEN` or `Tokenizer.NULL_CHARACTER_TOKEN`) into one token - but emitting a separate, new character token whenever this character token type changes. In these cases the end location (offset, line, col) of the preceding token comes out incorrect.

I'm still investigating, and trying to come up with a proper test case that can actually be included as a regression test. And maybe a fix. Unfortunately, the tests are organized a bit more complicated than they should be, IMHO...

**Comment by meisl** [comment]
Well, there is indeed a problem with the location-info extension, and it materializes already in tokens emitted from the tokenizer: #284.

However, since this (#282) is actually talking about the location info in *nodes* (from the parser, i.e. one level up), and, moreover, does *not* contain any of the problematic cases I have found there:
**I no longer think it's related**.
